File: libcore/src/main/java/com/mic/libcore/serial/SerialUtils.java
package com.mic.libcore.serial;

import android.serialport.SerialPort;
import com.mic.libcore.serial.callback.ISerialCallBack;
import com.mic.libcore.serial.constant.Baudrate;
import com.mic.libcore.serial.constant.DataType;
import com.mic.libcore.utils.KLog;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * 串口辅助
 * Created by onlykk on 2024-12-08
 */
public class SerialUtils {
    /**
     * 串口对象
     */
    private SerialPort mSerialPort;
    /**
     * 读取数据的线程
     */
    private Reader mReadThread;

    /**
     * 构造函数
     *
     * @param path     串口的物理地址
     * @param baudrate 串口的波特率
     */
    public SerialUtils(String path, Baudrate baudrate) {
        try {
            mSerialPort = SerialPort
                    .newBuilder(path, baudrate.getBaudrate()).build();
        } catch (IOException e) {
            KLog.e("串口打开失败：" + e.getMessage());
            mSerialPort = null;
        }
    }

    /**
     * 开始读取数据
     */
    public void startReadData(ISerialCallBack back, DataType dataType) {
        if (mSerialPort != null) {
            InputStream inputStream = mSerialPort.getInputStream();
            // 开始读取数据
            if (mReadThread == null) {
                mReadThread = new Reader(inputStream, back, dataType);
            }
            mReadThread.startReading();
        }
    }

    /**
     * 停止读取数据
     */
    public void stopReadData() {
        // 关闭读取线程
        if (mReadThread != null) {
            mReadThread.stopReading();
            mReadThread = null;
        }
        if (mSerialPort != null) {
            mSerialPort.tryClose();
        }
    }

    /**
     * 向串口上写数据
     *
     * @param data 显示的16进制的字符串
     */
    public synchronized void setData(byte[] data) {
        if (mSerialPort != null) {
            OutputStream outputStream = mSerialPort.getOutputStream();
            try {
                outputStream.write(data);
            } catch (Exception e) {
                KLog.i("写入数据异常：" + e.getMessage());
            }
        }
    }
}

File: support/support-serial/src/main/java/com/mic/guide/support/serial/SerialUtils.kt
package com.mic.guide.support.serial

import android.serialport.SerialPort
import com.mic.guide.lib.common.ByteUtils
import com.mic.guide.lib.log.Logger
import java.io.BufferedInputStream
import java.io.InputStream
import java.io.OutputStream

/**
 * 串口辅助（**合并自 libcore `SerialUtils` + `Reader`**，转 Kotlin）。
 *
 * 原 `Reader` 用 RxJava 轮询读取；这里改为一条守护读线程（与 [SerialManager] 同范式），
 * 去掉 RxJava 依赖。回调式 API，适合对接旧逻辑；需要 Flow 用 [SerialManager]。
 */
class SerialUtils(path: String, baudrate: Baudrate) {

    private var serialPort: SerialPort? = try {
        SerialPort.newBuilder(path, baudrate.value).build()
    } catch (e: Exception) {
        Logger.e("串口打开失败: ${e.message}", tag = "Serial")
        null
    }

    @Volatile
    private var reading = false
    private var readThread: Thread? = null

    /** 开始读取，按 [dataType] 把数据回调成字符串。 */
    fun startReadData(callback: SerialCallback, dataType: DataType) {
        val input: InputStream = serialPort?.inputStream ?: return
        if (reading) return
        reading = true
        readThread = Thread {
            val bis = BufferedInputStream(input)
            val buffer = ByteArray(512)
            while (reading) {
                try {
                    if (bis.available() > 0) {
                        val size = bis.read(buffer)
                        if (size > 0) {
                            val result = when (dataType) {
                                DataType.STRING -> String(buffer, 0, size)
                                DataType.HEX -> ByteUtils.bytes2HexStr(buffer, 0, size)
                            }
                            if (result.isNotEmpty()) callback.onResult(result)
                        }
                    } else {
                        Thread.sleep(10)
                    }
                } catch (e: Exception) {
                    if (reading) Logger.e("读取数据异常: ${e.message}", tag = "Serial")
                    break
                }
            }
        }.apply { isDaemon = true; start() }
    }

    fun stopReadData() {
        reading = false
        readThread?.interrupt()
        readThread = null
        runCatching { serialPort?.tryClose() }
        serialPort = null
    }

    /** 向串口写数据。 */
    @Synchronized
    fun setData(data: ByteArray) {
        val output: OutputStream = serialPort?.outputStream ?: return
        try {
            output.write(data)
        } catch (e: Exception) {
            Logger.i("写入数据异常: ${e.message}", tag = "Serial")
        }
    }
}
